lib/postgres.go

```diff
@@ -0,0 +1,1368 @@
+	FromUser        []byte
+	OtherUser       []byte
+	Type            NotificationType
+	Amount          uint64
+	PostHash        *BlockHash
+	Timestamp       uint64
+}
+
+type NotificationType uint8
+
+const (
+	NotificationUnknown NotificationType = iota
+	NotificationSendClout
+	NotificationLike
+	NotificationFollow
+	NotificationCoinPurchase
+	NotificationCoinTransfer
+	NotificationCoinDiamond
+	NotificationPostMention
+	NotificationPostReply
+	NotificationPostReclout
+)
+
+type Profile struct {
+	PKID                    *PKID `pg:",pk"`
+	PublicKey               *PublicKey
+	Username                string
+	Description             string
+	ProfilePic              []byte
+	CreatorBasisPoints      uint64
+	BitCloutLockedNanos     uint64
+	NumberOfHolders         uint64
+	CoinsInCirculationNanos uint64
+	CoinWatermarkNanos      uint64
+}
+
+type Post struct {
+	PostHash          *BlockHash `pg:",pk"`
+	PosterPublicKey   []byte
+	ParentPostHash    *BlockHash
+	Body              string
+	RecloutedPostHash *BlockHash
+	QuotedReclout     bool
+	Timestamp         uint64
+	Hidden            bool
+	LikeCount         uint64
+	RecloutCount      uint64
+	QuoteRecloutCount uint64
+	DiamondCount      uint64
+	CommentCount      uint64
+	Pinned            bool
+	ExtraData         map[string][]byte
+}
+
+func (post *Post) NewPostEntry() *PostEntry {
+	postEntry := &PostEntry{
+		PostHash:          post.PostHash,
+		PosterPublicKey:   post.PosterPublicKey,
+		Body:              []byte(post.Body),
+		RecloutedPostHash: post.RecloutedPostHash,
+		IsQuotedReclout:   post.QuotedReclout,
+		TimestampNanos:    post.Timestamp,
+		IsHidden:          post.Hidden,
+		LikeCount:         post.LikeCount,
+		RecloutCount:      post.RecloutCount,
+		QuoteRecloutCount: post.QuoteRecloutCount,
+		DiamondCount:      post.DiamondCount,
+		CommentCount:      post.CommentCount,
+		IsPinned:          post.Pinned,
+		PostExtraData:     post.ExtraData,
+	}
+
+	if post.ParentPostHash != nil {
+		postEntry.ParentStakeID = post.ParentPostHash.ToBytes()
+	}
+
+	return postEntry
+}
+
+func (post *Post) HasMedia() bool {
+	bodyJSONObj := BitCloutBodySchema{}
+	err := json.Unmarshal([]byte(post.Body), &bodyJSONObj)
+	// Return true if body json can be parsed and ImageUrls is not nil/non-empty or EmbedVideoUrl is not nil/non-empty
+	return (err == nil && len(bodyJSONObj.ImageURLs) > 0) || len(post.ExtraData["EmbedVideoURL"]) > 0
+}
+
+type Like struct {
+	LikerPublicKey []byte     `pg:",pk"`
+	LikedPostHash  *BlockHash `pg:",pk"`
+}
+
+func (like *Like) NewLikeEntry() *LikeEntry {
+	return &LikeEntry{
+		LikerPubKey:   like.LikerPublicKey,
+		LikedPostHash: like.LikedPostHash,
+	}
+}
+
+type Follow struct {
+	FollowerPKID *PKID `pg:",pk,type:bytea"`
+	FollowedPKID *PKID `pg:",pk,type:bytea"`
+}
+
+func (follow *Follow) NewFollowEntry() *FollowEntry {
+	return &FollowEntry{
+		FollowerPKID: follow.FollowerPKID,
+		FollowedPKID: follow.FollowedPKID,
+	}
+}
+
+type Diamond struct {
```

Comment:
You're probably going to hate this but it'd be ideal if we could change all these structs to be "PG[name]Entry".  So in this case, it would be "PGDiamondEntry".  The reason being, if you wanted to search for this specific struct in the codebase, searching for the word "diamond" would return a ton of erroneous matches.  Please discuss with @diamondhands0, if you think this is unnecessary... Sorry for the annoying nit!